Flask web app blog (#1664)
oktadeveloper/okta-blog: _source/_posts/2026-07-24-flask-oauth-web-app.md

Check out the complete source code on [GitHub](https://github.com/oktadev/flask-web-app-example) and get started without setting it up from scratch.

**Table of Contents**{: .hide }
* Table of Contents
{% include toc.md %}

## What you'll build

In this tutorial, you'll build a simple dashboard application and learn how to:

* Securely sign users in to view their profile information using Okta as the OpenID Connect (OIDC) identity provider.  
* Implement the Authorization Code flow with [Proof Key for Code Exchange (PKCE)]({% post_url 2025-07-28-express-oauth-pkce %}) to adhere to the latest OAuth 2.0 security best practices.  
* Add a feature for new users to self-register for an account directly from the sign-in page.  
* Enable authenticated users to interact with a protected API using OAuth 2.0 access tokens.

## Prerequisites

* [Python](https://www.python.org/) 3.14 or later and [pip](https://pypi.org/project/pip/) installed  
* An [Okta Integrator Free Plan](https://developer.okta.com/signup/)

## Create an app integration in the Okta Admin Console

> Before you begin, you'll need an Okta Integrator Free Plan account. To get one, sign up for an [Integrator account](https://developer.okta.com/login). Once you have an account, sign in to your [Integrator account](https://developer.okta.com/login). Next, in the Admin Console:

Reviewer: Good job adhering to the naming standards

> Before you begin, you'll need an Okta Integrator Free Plan account. To get one, sign up for an [Integrator account](https://developer.okta.com/login). Once you have an account, sign in to your [Integrator account](https://developer.okta.com/login). Next, in the Admin Console:

Reviewer: We have an embed for this. please use it

To include this scope in the tokens minted from Okta, follow these steps:

* In your Okta Admin Console, navigate to **Security** \> **API**.  
* Select your **Authorization Server** and go to the **Scopes** tab  
* Click the **Add Scope** button and enter `api:read-users` in the **Name** field.  
* Optionally, enter the **Description** and the **Display Phrase**  
* Click **Create**.

Now, when this scope is requested, Okta will include it in the tokens it issues.

Before creating the resource server, add an extra layer of security. You want your endpoint only to consume tokens intended for your server, preventing tokens issued for other services from being replayed or misused.

To do this, restrict the `aud` claim that Okta returns in the access token to your resource server. To achieve this:

* In your Okta Admin Console, navigate to **Security** \> **API**.  
* Select your **Authorization Server**. In the **Settings** tab, click **Edit**  
* Change the **Audience** field to [http://localhost:5001](http://localhost:5001) — this is where your resource server will be hosted

### Build a protected users API

> To keep things simple, you'll create a minimal resource server with a single route called `fetch-users` that returns a list of sample users.

Reviewer: ```suggestion
For this demo app, you'll create a minimal resource server with a single route called `fetch-users` that returns a list of sample users.
```

Once again, avoid insensitive words like simple if you can.

  }
 
  thead th {
    text-align: center;
    background-color: #e9f0fb;
    color: #1a3c72;
    font-weight: 700;
    padding: 0.75rem 1rem;
    border-bottom: 1px solid #d1d9e6;
    position: sticky;
    top: 0;
    z-index: 1;
  }
 
  td {
    text-align: center;
    padding: 0.75rem 1rem;
    border-bottom: 1px solid #e2e8f0;
    font-size: 1rem;
    word-wrap: break-word;
  }
```

### Test the protected API integration

Ready to see it all work? Make sure both servers are running: the Flask app (`flask run`) and the resource server (`flask --app resource-server.py run --port 5001`). Open a browser, go to http://localhost:5000, sign in, and then click Fetch Users. A table should appear with a list of sample users returned from your resource server. If you see the user data, your integration is a success\!

{% img blog/flask-oauth-web-app/flask-app-fetch-users.jpg alt:"Flask web app displaying a users table with email, first name, and last name columns after successfully fetching data from the protected API" width:"800" %}{: .center-image }

Your Flask application now has secure user authentication and can use an access token to interact with a protected API.

> ## Learn more

Reviewer: can you make this a better section header and work in your keywords again?
